File: Implementierung/Backend/src/main/java/edu/kit/informatik/service/CategoryService.java
package edu.kit.informatik.service;


import edu.kit.informatik.dto.mapper.interactions.CategoryMapper;
import edu.kit.informatik.dto.mapper.interactions.RatedCategoryMapper;
import edu.kit.informatik.dto.userdata.interactions.CategoryDto;
import edu.kit.informatik.dto.userdata.interactions.RatedCategoryDto;
import edu.kit.informatik.exceptions.EntityNotFoundException;
import edu.kit.informatik.model.userdata.interactions.Category;
import edu.kit.informatik.model.userdata.interactions.RatedCategory;
import edu.kit.informatik.repositories.CategoryBaseRepository;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

/**
 * Service für {@link Category Kategorien}
 *
 * @author ugqbo
 * @version 0.1
 */

@Component
@EnableAutoConfiguration
public class CategoryService extends BaseService<Category, RatedCategoryDto, CategoryDto> {

    private static final String ID_ATTRIBUTE = "userId";

    private final CategoryBaseRepository<Category> categoryBaseRepository;
    private final RatedCategoryMapper ratedCategoryMapper;

    /**
     * Konstruktor zum Erstellen eines Objektes der Klasse
     *
     * @param categoryBaseRepository {@link CategoryBaseRepository}
     * @param categoryMapper         {@link CategoryMapper}
     * @param ratedCategoryMapper    {@link RatedCategoryMapper}
     */
    public CategoryService(CategoryBaseRepository<Category> categoryBaseRepository,
                           CategoryMapper categoryMapper, RatedCategoryMapper ratedCategoryMapper) {
        super(categoryMapper);
        this.categoryBaseRepository = categoryBaseRepository;
        this.ratedCategoryMapper = ratedCategoryMapper;
    }

    @Override
    public CategoryDto add(RatedCategoryDto categoryDto, Authentication authentication) {
        super.checkAuthorization(authentication, categoryDto.getUserId());
        if (categoryDto.getQuality() == null) {
            Category newCategory = this.categoryBaseRepository.save(this.mapper.dtoToModel(categoryDto));
            return this.mapper.modelToDto(newCategory);
        } else {
            RatedCategory newCategory = this.categoryBaseRepository.save(
                                                this.ratedCategoryMapper.dtoToModel(categoryDto));
            return this.ratedCategoryMapper.modelToDto(newCategory);
        }

    }

    @Transactional
    @Override
    public CategoryDto update(RatedCategoryDto categoryDto, Authentication authentication) {
        super.checkAuthorization(authentication, categoryDto.getUserId());
        Optional<Category> repositoryCategoryOptional = this.categoryBaseRepository
                                                                .findCategoryById(categoryDto.getId());

        Category repositoryCategory = repositoryCategoryOptional.orElseThrow(() ->
                                                    new EntityNotFoundException(Category.class, categoryDto.getId()));
        Category newCategory = this.mapper.dtoToModel(categoryDto);

        if (!newCategory.getName().equals(repositoryCategory.getName())) {
            repositoryCategory.setName(newCategory.getName());
        }

        return mapper.modelToDto(repositoryCategory);
    }

    @Override
    public CategoryDto getById(String id, Authentication authentication) {
        Optional<Category> categoryOptional = this.categoryBaseRepository.findCategoryById(id);
        Category category = categoryOptional.orElseThrow(() -> new EntityNotFoundException(Category.class, id));
        super.checkAuthorization(authentication, category.getUser().getId());

        return this.mapper.modelToDto(category);
    }

    @Override
    public List<CategoryDto> getAll(Authentication authentication) {
        JwtAuthenticationToken jAT = (JwtAuthenticationToken) authentication;

        return this.mapper.modelToDto(this.categoryBaseRepository.findCategoryByUserId(
                                                                jAT.getTokenAttributes().get(ID_ATTRIBUTE).toString()));
    }

    @Override
    public String delete(String id, Authentication authentication) {
        Optional<Category> categoryOptional = this.categoryBaseRepository.findCategoryById(id);
        Category category = categoryOptional.orElseThrow(() -> new EntityNotFoundException(Category.class, id));
        super.checkAuthorization(authentication, category.getUser().getId());

        return delete(category);
    }

    /**
     * Methode zum Löschen einer Category
     * --> dient für andere Services
     * @param category {@link Category}
     * @return Id der {@link Category}
     */
    protected String delete(Category category) {
        this.categoryBaseRepository.deleteById(category.getId());
        return category.getId();
    }
}

File: Implementierung/frontend/paschi/src/service/CategoryService.ts
import {BASE_URL, BaseService} from "@/service/BaseService";
import { Category } from "@/model/userdata/interactions/Category";
import { CategoryDto } from "@/dto/userdata/interactions/CategoryDto";
import { CategoryMapper } from "@/dto/mapper/interactions/CategoryMapper";
import axios, { AxiosResponse } from "axios";
import { useUserStore } from "@/store/UserStore";

const CATEGORY_BASE_URL: string = BASE_URL + "/api/category";

export class CategoryService extends BaseService<Category, CategoryDto> {

  private static categoryService: CategoryService = new CategoryService();

  private constructor() {
    super(CategoryMapper.getMapper());
  }

  static getService(): CategoryService {
    return this.categoryService;
  }

  async add(category: Category) {
    const token = useUserStore().getUser()?.token;
    const categoryDto = this.getMapper().modelToDto(category);
    await axios
      .post(CATEGORY_BASE_URL, categoryDto, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then((response: AxiosResponse<CategoryDto>) => {
        category.setId = response.data.id;
      })
      .catch((error) => {
        console.log(error);
      });
  }

  async update(category: Category) {
    const token = useUserStore().getUser()?.token;
    const categoryDto = this.getMapper().modelToDto(category);
    await axios
      .put(CATEGORY_BASE_URL, categoryDto, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .catch((error) => {
        console.log(error);
      });
  }

  async getById(id: string): Promise<Category | undefined> {
    const token = useUserStore().getUser()?.token;
    let category;
    await axios
      .get(CATEGORY_BASE_URL + `/${id}`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then(async (response: AxiosResponse<CategoryDto>) => {
        category = await this.getMapper().dtoToModel(response.data);
      })
      .catch((error) => {
        console.log(error);
      });

    return category;
  }

  async getAll(): Promise<Category[]> {
    const token = useUserStore().getUser()?.token;
    const categories: Category[] = [];
    await axios
      .get(CATEGORY_BASE_URL, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then(async (response: AxiosResponse<CategoryDto[]>) => {
        for (const categoryDto of response.data) {
          categories.push(await this.getMapper().dtoToModel(categoryDto));
        }
      })
      .catch((error) => {
        console.log(error);
      });

    return categories;
  }

  async delete(id: string) {
    const token = useUserStore().getUser()?.token;
    await axios
      .delete(CATEGORY_BASE_URL, {
        params: {
          id,
        },
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .catch((error) => {
        console.log(error);
      });
  }
}
